subprojects/core/src/main/java/org/gradle/execution/taskgraph/DefaultTaskExecutionPlan.java

```diff
@@ -475,65 +464,93 @@ public void useFailureHandler(TaskFailureHandler handler) {
+                                        recordTaskStarted(taskInfo);
+                                        canExecute.set(true);
+                                    } else {
+                                        taskInfo.skipExecution();
+                                    }
+                                    return FINISHED;
+                                } else {
+                                    return FAILED;
+                                }
+                            }
+                        });
+
+                        if (selected.get() != null) {
+                            break;
+                        }
                     }
                 }
-                if (allTasksComplete) {
-                    return null;
-                }
-                if (nextMatching == null) {
-                    try {
-                        condition.await();
-                    } catch (InterruptedException e) {
-                        throw new RuntimeException(e);
-                    }
+
+                if (selected.get() == null && !allTasksComplete()) {
+                    return RETRY;
                 } else {
-                    if (nextMatching.allDependenciesSuccessful()) {
-                        nextMatching.startExecution();
-                        recordTaskStarted(nextMatching);
-                        return nextMatching;
-                    } else {
-                        nextMatching.skipExecution();
-                        condition.signalAll();
-                    }
+                    return FINISHED;
+                }
+            }
+        });
+
+
+        if (selected.get() != null) {
+            try {
+                if (canExecute.get()) {
+                    taskExecution.execute(selected.get());
                 }
+            } finally {
+                coordinationService.withStateLock(new Transformer<ResourceLockState.Disposition, ResourceLockState>() {
+                    @Override
+                    public ResourceLockState.Disposition transform(ResourceLockState resourceLockState) {
+                        TaskInfo taskInfo = selected.get();
+                        ResourceLock projectLock = getProjectLock(taskInfo);
+                        projectLock.unlock();
+                        workerLease.unlock();
+                        return FINISHED;
+                    }
+                });
             }
-        } finally {
-            lock.unlock();
         }
+
+        // If all tasks are complete, we're done
+        return !allTasksComplete();
+    }
+
+    private ResourceLock getProjectLock(TaskInfo taskInfo) {
+        Project project = taskInfo.getTask().getProject();
+        String gradlePath = ((GradleInternal) project.getGradle()).getIdentityPath().toString();
+        String projectPath = ((ProjectInternal) project).getIdentityPath().toString();
+        return workerLeaseService.getProjectLock(gradlePath, projectPath);
```

Comment:
Per my comment above, passing in a `Project` would require that class to be visible, which would pull these services down into core.  I think this is a reasonable trade off in order to keep them in base services.